features/lorawan/lorastack/phy/LoRaPHY.cpp

```diff
@@ -388,23 +389,49 @@ uint8_t LoRaPHY::verify_link_ADR_req(verify_adr_params_t *verify_params,
     return status;
 }
 
-double LoRaPHY::compute_symb_timeout_lora(uint8_t phy_dr, uint32_t bandwidth)
+float LoRaPHY::compute_symb_timeout_lora(uint8_t phy_dr, uint32_t bandwidth)
 {
-    return ((double)(1 << phy_dr) / (double) bandwidth) * 1000;
+    // in milliseconds
+    return ((float)(1 << phy_dr) / (float) bandwidth * 1000);
 }
 
-double LoRaPHY::compute_symb_timeout_fsk(uint8_t phy_dr)
+float LoRaPHY::compute_symb_timeout_fsk(uint8_t phy_dr)
 {
-    return (8.0 / (double) phy_dr); // 1 symbol equals 1 byte
+    return (8.0 / (float) phy_dr); // 1 symbol equals 1 byte
 }
 
-void LoRaPHY::get_rx_window_params(double t_symb, uint8_t min_rx_symb,
-                                   uint32_t rx_error, uint32_t wakeup_time,
-                                   uint32_t *window_timeout, int32_t *window_offset)
+
+void LoRaPHY::get_rx_window_params(float t_symb, uint8_t min_rx_symb,
+                                   float error_fudge, float wakeup_time,
+                                   uint32_t *window_length, int32_t *window_offset,
+                                   uint8_t phy_dr)
 {
-    // Computed number of symbols
-    *window_timeout = MAX((uint32_t) ceil(((2 * min_rx_symb - 8) * t_symb + 2 * rx_error) / t_symb), min_rx_symb);
-    *window_offset = (int32_t) ceil((4.0 * t_symb) - ((*window_timeout * t_symb) / 2.0) - wakeup_time);
+    float min_rx_symbol_overlap_required = float (min_rx_symb);
+    float target_rx_window_offset;
+    float window_len_in_ms;
+
+    if (phy_params.fsk_supported && phy_dr == phy_params.max_rx_datarate) {
+        min_rx_symbol_overlap_required = MAX_PREAMBLE_LENGTH;
+    }
+
+    // We wish to be as close as possible to the actual start of data, i.e.,
+    // we are interested in the preamble symbols which are at the tail of the
+    // preamble sequence.
+    target_rx_window_offset = (MAX_PREAMBLE_LENGTH - min_rx_symbol_overlap_required) * t_symb; //in ms
+
+    // Actual window offset in ms in response to timing error fudge factor and
+    // radio wakeup/turned around time.
+    *window_offset = floor(target_rx_window_offset - error_fudge - wakeup_time);
```

Comment:
Question for @c1728p9 - does/will the sleep code compensate for the 10ms or whatever deep sleep latency, if a long sleep is requested?

Eg if `wait(1000)` or `call_in(1000)`, do we get a wake from deep sleep at 990 then wake from shallow at 1000, so we hit 1000 precisely, or do we just schedule an interrupt at 1000 so we potentially wake at 1010?

This code in particular is trying to hit a 1 second delay for a receive window after a transmit with millisecond precision. Can that be achieved without holding a deep sleep lock?

Reply:
Great, that's what I hoped. I knew you'd been looking at the compensation, but couldn't recall if it was just covering "teeny" sleeps or reached out for long targets.

In this case, it's via an `EventQueue::call_in`, so that's a timed semaphore wake, so we're good.

This "sleep latency compensation only via the OS" thing feels like it should be documented somewhere, but not sure where to encourage people to use OS mechanisms rather than `LowPowerTimer`s. Some sort of knowledge base article about deep sleep/tickless or something. Or maybe even a note on `LowPowerTimer`, which is what people might be thinking of using with deep sleep active.